drivers/at_cmd/at_cmd.c

```diff
@@ -172,75 +146,42 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
 
 		payload_len = get_return_code(buf, &ret);
 
-		if (ret.state != AT_CMD_NOTIFICATION) {
-			if ((response_buf_len > 0) &&
-			    (response_buf != NULL)) {
-				if (response_buf_len > payload_len) {
-					memcpy(response_buf,
-						buf, payload_len);
-				} else {
-					LOG_ERR("Response buffer not large "
-						"enough");
-
-					ret.code  = -EMSGSIZE;
-				}
-
-				callback         = false;
-				response_buf_len = 0;
-				response_buf     = NULL;
+		/* Handle AT notification separately. */
+		if (ret.state == AT_CMD_NOTIFICATION) {
+			if (notification_handler != NULL) {
+				notification_handler(buf);
 			}
+
+			continue;
 		}
 
 		if (payload_len == 0) {
 			goto next;
 		}
 
-		if (callback) {
-			at_cmd_handler_t tmp_callback;
-
-			if (ret.state == AT_CMD_NOTIFICATION) {
-				tmp_callback = notification_handler;
+		/* Handle AT command response. */
+		if ((response_buf_len > 0) && (response_buf != NULL)) {
+			/* If buffers were provided, copy response. */
+			if (response_buf_len > payload_len) {
+				memcpy(response_buf, buf, payload_len);
 			} else {
-				tmp_callback = current_cmd_handler;
+				LOG_ERR("Response buffer not large enough");
+				ret.code  = -EMSGSIZE;
 			}
-
-			if (!tmp_callback) {
-				goto next;
-			}
-
-			struct callback_work_item *item =
-			k_malloc(sizeof(struct callback_work_item));
-
-			if (!item) {
-				LOG_DBG("Failed to allocate work item");
-
-				ret.code  = -ENOMEM;
-
-				goto next;
-			}
-
-			item->callback = tmp_callback;
-
-			k_work_init(&item->work, callback_worker);
-
-			memcpy(item->data, buf, payload_len);
-
-			k_work_submit(&item->work);
+		} else if (current_cmd_handler != NULL) {
+			/* Otherwise, call the callback. */
+			current_cmd_handler(buf);
```

Comment:
I did this in my implementation, but it may result in a deadlock. A likely scenario is that want to send a new AT command in this callback you will be stuck on the at_cmd_write_* call waiting for the semaphore because it's the same thread that is keeping it. That is exactly why I started offload the callback execution to the work queue.

Reply:
Ok, I see now why was it implemented this way. 

Still, I'm not happy about using a work queue for this use-case. `at_cmd_write_*` calls are blocking calls, and these by principle should be avoided in a system work queue.

To avoid deadlocks, we could move the cmd_handler call back to the caller thread (by calling it from within `at_cmd_write_*` function, not the RX thread). Of course, mutex would need to be freed before, but that's not the case - it has already served its job. I need to think about it.

Reply:
BTW is there an existing use-case for the scenario you describe? I've searched repos that could've used this API and haven't found any? So perhaps we could simply add a limitation in the docs, that this API should not be called from within callback context and stick to a more memory efficient implementation with a single buffer?

Reply:
I don't think there is such scenario. In the simplest case, the application will call the `at_cmd_write` without supplying a callback, thus blocking. Once the call has returned, the application can send more AT commands based on the previous response. This is the simplest way to do it and how it is done currently in most/all places.

I think it's acceptable that if a callback is provided, the application shouldn't issue more commands in the callback, but instead schedule any non-blocking processing to be done in the system workqueue.

Reply:
The lwm2m library do this, when it receive an SMS it will ACK the SMS from the notification callback.

Reply:
@rlubos Even if you move the calling of the callback to the caller thread you will still need to copy the data to a different buffer, so it will not solve the heap issue. 

To day the callback_work_item has buffer with an statick size of 2700 (default configuratiion) bytes. I don't remember doing this, but it was probably done during the review process and side effects where not considered. Anyway the simples fix would be to do two allocation, one for the callback_work_item header and one for the buffer containing the response/notification payload. In most cases this would result in 128 bytes allocated on the heap pr work item, not 2700 bytes. 

A different solution would be to use the memory pool or memory slabs APIs which uses a static buffer. Or I have also done some trials with the ringbuf, but it seems to be a bad fit.

Reply:
Memory pools seems to be a nice fit, it's a two liner to change the code to use a static memory pool instead

Reply:
I did the changes here
https://github.com/evenl/fw-nrfconnect-nrf/tree/at_cmd_update_2

I also though that I could re-purpose the message queue currently sending status information back to the caller to also include the callback reference, so that it could be called within the at_cmd_write_* functions. And only call notification callbacks from within the AT thread.

Reply:
Ok, I got confused about the line we discuss, I thought the use-case is `at_cmd_write_with_callback` (which apparently isn't used anywhere), not the notification handler. Seeing how it's used in LWM2M lib, indeed there's no way to make it work in a single-buffer scenario.

> Memory pools seems to be a nice fit

Zephyr heap is built on top of memory poll, we might hit the same issue here. I would consider memory slab for this use-case - a set of fixed-sized buffers:
https://docs.zephyrproject.org/latest/reference/kernel/memory/slabs.html
The number of buffers could be configurable with Kconfig (with 2 at minimum to avoid deadlocks). Size of a single block should match the size of workqueue item.

> I changed the default message size down to 256 bytes

But what about the PR's which increased the message size? I hope it was done for a purpose, not to make our lives harder. I'd keep it as it is for now and possibly discuss the trim down with authors of those PRs.

I can rework the PR to use mem slabs on Monday.

Reply:
Ok, yes, that is true, removing at_cmd_write_with_callback is actually a solution to the problem. A call to at_cmd_write from the callback in this case will actually work because this thread has not taken the semaphore already, so if it's blocked it will be by a different thread. Then I think your patch is already valid (except, that it doesn't remove at_cmd_write_with_callback)

So then is the question if at_cmd_write_with_callback is not an use-case in the long run.

Reply:
Sorry, my bad, what I said is not correct. You will still be stuck waiting for the result to come back on the message queue. So we need this callback to be executed by a different thread.